Add environment_variables.txt file documentation (#13669)
mlflow/mlflow: docs/source/models.rst

^^^^^^^^^^^^

All of the flavors that a particular model supports are defined in its ``MLmodel`` file in YAML

> format. For example, running ``python examples/sklearn_logistic_regression/train.py`` from `MLflow repo <https://github.com/mlflow/mlflow>`_

Reviewer: ```suggestion
format. For example, running ``python examples/sklearn_logistic_regression/train.py`` from `MLflow repo <https://github.com/mlflow/mlflow/blob/master/examples/sklearn_logistic_regression/train.py>`_
```

can we link to the example?

::

>     # Run `python examples/sklearn_logistic_regression/train.py` from mlflow directory

Reviewer 2: ```suggestion
    # Running `python examples/sklearn_logistic_regression/train.py` will create the following files under the `model` directory.
```

>     model/

Reviewer: I'd prefer the original description. If I was a user, I would wonder "where does examples/sklearn_logistic_regression/train.py come from?".

    ├── MLmodel
    ├── model.pkl
    ├── conda.yaml
    ├── python_env.yaml
    ├── requirements.txt
    ├── input_example.json

>     ├── serving_input_example.json

Reviewer: `input_example.json` and `serving_input_example.json` are also optional, right?

>     ├── serving_input_example.json

Reviewer 2: Let's comment these are only created when an input example is passed.

>     └── environment_variables.txt (optional, only existing when environment variables are used)

Reviewer: ```suggestion
    └── environment_variables.txt (optional, only logged when environment variables are used during model inference)
```

We should be more precise.

These files can then be used to reinstall dependencies using ``conda`` or ``virtualenv`` with ``pip``. Please see 
:ref:`How MLflow Model Records Dependencies <how-mlflow-records-dependencies>` for more details about these files.

> If the model input example is provided when logging the model, additional files including ``input_example.json`` and ``serving_input_example.json`` are logged.

Reviewer 2: ```suggestion
If a model input example is provided when logging the model, two additional files ``input_example.json`` and ``serving_input_example.json`` are logged.
```

    model/dependencies

Environment variables file
^^^^^^^^^^^^^^^^^^^^^^^^^^

> To track environment variables that are used during model inference, MLflow generates ``environment_variables.txt`` file when logging a model.

Reviewer: ```suggestion
MLflow generates logs the environment variables used during model inference in ``environment_variables.txt`` file when logging a model.
```

> It **only contains names** of the environment variables that are used during model inference, **values are not stored**.

Reviewer 2: nit: Can we make this `.. attention:`?

> Currently we only log the name if it includes any keywords in the following set:

Reviewer: can we replace we with mlflow?

> Currently MLflow only logs the name if it includes any keywords in the following set:

Reviewer 2: ```suggestion
Currently MLflow only logs the environment variables whose name contains any of the following keywords:
```

.. code-block:: python

    RECORD_ENV_VAR_ALLOWLIST = {
        # api key related

>         "API_KEY",

Reviewer: ```suggestion
        "API_KEY",  # e.g. OPENAI_API_KEY
```

        "API_TOKEN",
        # databricks auth related
        "DATABRICKS_HOST",
        "DATABRICKS_USERNAME",
        "DATABRICKS_PASSWORD",
        "DATABRICKS_TOKEN",
        "DATABRICKS_INSECURE",
        "DATABRICKS_CLIENT_ID",
        "DATABRICKS_CLIENT_SECRET",
        "_DATABRICKS_WORKSPACE_HOST",
        "_DATABRICKS_WORKSPACE_ID",
        # dbconnect related
        "SPARK_REMOTE",
        "SPARK_LOCAL_REMOTE",
    }

> 

Reviewer: Can add an example model that uses environment variables?

.. attention::

>     **Check the environment_variables.txt file before deploying the model to a serving endpoint**, as they might need 
>     to be set in the serving environment. Follow `this guidance <https://docs.databricks.com/en/machine-learning/model-serving/store-env-variable-model-serving.html#add-plain-text-environment-variables>`_
>     to set environment variables on a databricks serving endpoint. **Note that it is not guaranteed that all environment
>     variables tracked in the file are required for model inference.**

Reviewer 2: ```suggestion
    Before you deploy a model to a serving endpoint, **review the ``environment_variables.txt`` file** to ensure all necessary environment variables for model inference are set. Note that **not all environment 
    variables listed in the file are always required for model inference.** For detailed instructions on setting environment variables on a databricks serving endpoint, refer to `this guidance <https://docs.databricks.com/en/machine-learning/model-serving/store-env-variable-model-serving.html#add-plain-text-environment-variables>`_.
```

.. note::

>     To disable logging environment variable names in model logging, set the environment variable 

Reviewer 2: ```suggestion
    To disable this feature, set the environment variable 
```

        # databricks auth related
        "DATABRICKS_HOST",
        "DATABRICKS_USERNAME",
        "DATABRICKS_PASSWORD",
        "DATABRICKS_TOKEN",
        "DATABRICKS_INSECURE",
        "DATABRICKS_CLIENT_ID",
        "DATABRICKS_CLIENT_SECRET",
        "_DATABRICKS_WORKSPACE_HOST",
        "_DATABRICKS_WORKSPACE_ID",
        # dbconnect related
        "SPARK_REMOTE",
        "SPARK_LOCAL_REMOTE",
    }

Example of a pyfunc model that uses environment variables:

.. code-block:: python

    import mlflow
    import os

    os.environ["TEST_API_KEY"] = "test_api_key"

    class MyModel(mlflow.pyfunc.PythonModel):
        def predict(self, context, model_input, params=None):
            if os.environ.get("TEST_API_KEY"):
                return model_input
            raise Exception("API key not found")

    with mlflow.start_run():
        model_info = mlflow.pyfunc.log_model(
            "model", python_model=MyModel(), input_example="data"
        )

Environment variable `TEST_API_KEY` is logged in the environment_variables.txt file

.. figure:: ./_static/images/models/environment_variable_file_sample.png
    :alt: The environment variables txt file example
    :width: 80%

>     :align: center

Reviewer: Can we use a code block instead?
